### PR TITLE
feat: upload final cmo (FPLA-1915)

### PR DESCRIPTION
**Pull request by lewisbirks**
### JIRA link (if applicable) ###

[FPLA-1915](https://tools.hmcts.net/jira/browse/FPLA-1915)

### Change description ###

Introduced new interim CMO functionality, deprecated previous draft CMO event.
Most deprecations have been moved to another PR to be done at a later date however there might be some deprecations involved in this PR that can be moved to another PR.

**Does this PR introduce a breaking change?** (check one with "x")

```
[] Yes
[x] No
```

### CODE REVIEW

File: ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json [hunk 1]
```diff
@@ -131,5 +131,12 @@
     "CaseEventID": "populateCase-PREPARE_FOR_HEARING",
     "UserRole": "caseworker-publiclaw-solicitor",
     "CRUD": "R"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "uploadCMO",
```
[the remark this answers is not in the document]
Comment (akomisarek) [reply ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json]: I believe this was fixed in subsequent commits.

File: ccd-definition/CaseTypeTab/CareSupervision/3_DraftOrders.json [hunk 2]
```diff
@@ -38,7 +38,17 @@
...
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "Channel": "CaseWorker",
+    "TabID": "DraftOrdersTab",
+    "TabLabel": "Draft orders",
+    "TabDisplayOrder": 3,
+    "CaseFieldID": "draftUploadedCMOs",
+    "TabFieldDisplayOrder": 3,
+    "TabShowCondition": "standardDirectionOrder.orderStatus!=\"SEALED\" OR caseManagementOrder!=\"\" OR sharedDraftCMODocument!=\"\" OR cmoToAction!=\"\" OR draftUploadedCMOs!=\"\""
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/CaseTypeTab/CareSupervision/3_DraftOrders.json:52]: I remember there being issues with the condition being on the first element of the tab when we were first doing CMO stuff

File: ccd-definition/AuthorisationCaseEvent/CareSupervision/DEPRECATED.json [hunk 3]
```diff
@@ -460,5 +460,19 @@
...
+    "CaseEventID": "draftCMO",
+    "UserRole": "caseworker-publiclaw-solicitor",
+    "CRUD": "R"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "draftCMO",
+    "UserRole": "[LASOLICITOR]",
+    "CRUD": "R"
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/AuthorisationCaseEvent/CareSupervision/DEPRECATED.json:476]: There is that one case that has used this event though isn't there? Are we happy removing their visibility to this?

File: ccd-definition/ComplexTypes/CareSupervision/CaseManagementOrder/CaseManagementOrder.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
+[
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/ComplexTypes/CareSupervision/CaseManagementOrder/CaseManagementOrder.json:1]: I was considering this, could make some of the submitted event code a bit simpler will have a look into it

[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/ComplexTypes/CareSupervision/CaseManagementOrder/CaseManagementOrder.json:1]: @joez-f Would this provide any use for you if I added it?

File: ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json]: All roles now have read, only LASOLICITOR has create and update in addition to that

File: ccd-definition/ComplexTypes/CareSupervision/CaseManagementOrder/CaseManagementOrder.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/ComplexTypes/CareSupervision/CaseManagementOrder/CaseManagementOrder.json:1]: Seems that it won't make anything easier for me but it might have some future uses so could do.

File: ccd-definition/AuthorisationCaseEvent/CareSupervision/DEPRECATED.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (akomisarek) [reply ccd-definition/AuthorisationCaseEvent/CareSupervision/DEPRECATED.json:476]: Probably, business question, but I would say yes,I don't think it was even properly used, more like testing, and we wouldn't break anything - the data would be there.

File: bin/build-release-ccd-definition.sh [hunk 5]
```diff
@@ -12,7 +12,7 @@ fi
 if [ ${environment} == "prod" ]; then
   excludedFilenamePatterns="-e UserProfile.json,*-nonprod.json"
 else
-  excludedFilenamePatterns="-e UserProfile.json"
+  excludedFilenamePatterns="-e UserProfile.json,*-prod.json"
```
Comment (Jamesnelson117) [inline bin/build-release-ccd-definition.sh:15]: @joez-f will you need this to? Happy to make a separate PR

Comment (lewisbirks) [reply bin/build-release-ccd-definition.sh:15]: Can just do
```
git checkout origin/FPLA-1990-View-Application-In-Open-State bin/import-ccd-definition.sh
git checkout origin/FPLA-1990-View-Application-In-Open-State bin/build-release-ccd-definition.sh
```
if you can't be bothered to

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        // update judge and hearing labels
+        Object dynamicList = caseData.getPastHearingList();
+        List<Element<HearingBooking>> hearings = cmoService.getHearingsWithoutCMO(caseData.getPastHearings(),
+            caseData.getDraftUploadedCMOs());
+        UUID selectedHearing = cmoService.getSelectedHearingId(dynamicList, hearings);
+        data.putAll(cmoService.getJudgeAndHearingDetails(selectedHearing, hearings));
+
+        if (!(dynamicList instanceof DynamicList)) {
+            // reconstruct dynamic list
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java:72]: Ah the dynamiclist... Do you think it's worth marking with the same Todo as in another place?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    public AboutToStartOrSubmitCallbackResponse handleLabelMidEvent(@RequestBody CallbackRequest request) {
+        Map<String, Object> data = request.getCaseDetails().getData();
+        CaseData caseData = mapper.convertValue(data, CaseData.class);
+
+        // update judge and hearing labels
+        Object dynamicList = caseData.getPastHearingList();
+        List<Element<HearingBooking>> hearings = cmoService.getHearingsWithoutCMO(caseData.getPastHearings(),
+            caseData.getDraftUploadedCMOs());
+        UUID selectedHearing = cmoService.getSelectedHearingId(dynamicList, hearings);
+        data.putAll(cmoService.getJudgeAndHearingDetails(selectedHearing, hearings));
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java:69]: What do you think about abstracting this inside the service itself? This contain a bit of logic and orchestration which ideally, shouldn't be there.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        List<Element<CaseManagementOrder>> draftCMOs = caseData.getDraftUploadedCMOs();
+        List<Element<CaseManagementOrder>> draftCMOsBefore = caseDataBefore.getDraftUploadedCMOs();
+        draftCMOs.removeAll(draftCMOsBefore);
+
+        if (draftCMOs.size() == 1) {
+            List<Element<HearingBooking>> hearings = caseData.getHearingDetails();
+            List<Element<HearingBooking>> hearingsBefore = caseDataBefore.getHearingDetails();
+            hearings.removeAll(hearingsBefore);
+
+            publisher.publishEvent(new CMOReadyToSealEvent(request, hearings.get(0).getValue()));
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java:143]: CMOReadyToSealEvent it's bad "Event" name. 

1. I would avoid Event in the name (I know we have in many places but it's old :P). 
2. This place of code shouldn't decide if something is ready or not, it should more state the fact. 

I would suggest more something like "NewCMOUploaded"?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    @PostMapping("/submitted")
+    public void handelSubmitted(@RequestBody CallbackRequest request) {
+        CaseData caseDataBefore = mapper.convertValue(request.getCaseDetailsBefore().getData(), CaseData.class);
+        CaseData caseData = mapper.convertValue(request.getCaseDetails().getData(), CaseData.class);
+
+        List<Element<CaseManagementOrder>> draftCMOs = caseData.getDraftUploadedCMOs();
+        List<Element<CaseManagementOrder>> draftCMOsBefore = caseDataBefore.getDraftUploadedCMOs();
+        draftCMOs.removeAll(draftCMOsBefore);
+
+        if (draftCMOs.size() == 1) {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java:138]: This is a bit logic, could it be abstracted to the service? It could be probably something like if(cmoService.anyNewCmoUploaded()) or something like that?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+            if (uuid != null) {
+                // overwrite old draft CMO
+                int index = -1;
+                for (int i = 0; i < draftCMOs.size(); i++) {
+                    if (draftCMOs.get(i).getId().equals(uuid)) {
+                        index = i;
+                        break;
+                    }
+                }
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java:110]: I would argue it's too much logic here.

The reason is why we want to decouple logic from controller is - in modern applications, you can have often multiple entry points to the system (i.e. REST API / Messaging system as best example), it's best if you can easily invoke the same processing from two different places.

I know we want be there here, but it's always to have the good principles. And possibly you can avoid some controller tests by pushing this further down, as this should be tested by service tests.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/events/cmo/CMOReadyToSealEvent.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import uk.gov.hmcts.reform.ccd.client.model.CallbackRequest;
+import uk.gov.hmcts.reform.fpl.events.CallbackEvent;
+import uk.gov.hmcts.reform.fpl.model.HearingBooking;
+
+@Getter
+@EqualsAndHashCode(callSuper = true)
+public class CMOReadyToSealEvent extends CallbackEvent {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/events/cmo/CMOReadyToSealEvent.java:11]: Already menttioned the event name.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/HearingBooking.java [hunk 12]
```diff
@@ -43,4 +50,13 @@ public boolean hasDatesOnSameDay() {
     public boolean startsAfterToday() {
         return startDate.isAfter(ZonedDateTime.now(ZoneId.of("Europe/London")).toLocalDateTime());
     }
+
+    public boolean hasCMOAssociation() {
+        return caseManagementOrderId != null;
+    }
+
+    public String toLabel(String dateFormat) {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/HearingBooking.java:58]: I don't massively mind it here, but I don't think it's an ideal place - the reason is different "pages" might expect different representation of hearingBooking label. But it's fine for the starters.

We have got Welsh language looming as well, having English text scattered through the codebase is far from ideal.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/CaseManagementOrder.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+@Data
+@Builder
+public class CaseManagementOrder {
+    private DocumentReference order;
+    private String hearing;
+    private LocalDate dateSent;
+    private CMOStatus status;
+
+    public static CaseManagementOrder from(DocumentReference order, HearingBooking hearing, LocalDate date) {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/CaseManagementOrder.java:22]: Nice handy utility factory method, I like it 👍

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/cmo/CMOReadyToSealEventHandlerTest.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    private NotificationService notificationService;
+
+    @Mock
+    private HmctsAdminNotificationHandler adminNotificationHandler;
+
+    @Mock
+    private CMOReadyToSealContentProvider contentProvider;
+
+    @Mock
+    private ObjectMapper mapper;
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/cmo/CMOReadyToSealEventHandlerTest.java:53]: I personally would never mock ObjectMapper. It's an easy object to prepare "real thing" and in my opinion lead to better tests in general

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/cmo/CMOReadyToSealEventHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    private static final String HMCTS_JUDGE_EMAIL = "judge@hmcts.gov.uk";
+
+    @Mock
+    private NotificationService notificationService;
+
+    @Mock
+    private HmctsAdminNotificationHandler adminNotificationHandler;
+
+    @Mock
+    private CMOReadyToSealContentProvider contentProvider;
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/cmo/CMOReadyToSealEventHandlerTest.java:50]: I think NotificationServicce is the only really one thing I would mock. This might be unpopular opinion, but I don't like religious "unit" tests and mocking every possible dependency - it makes tests like that extremely brittle while refactoring. Would you expect the code handler to fail if I move the code a bit, but the functionality would still work? 
HmctsAdminNotificationHandler - I know it's not your code, probably remainder of some old logic, but it's not even handler, it just gives address from config. 

Of course the above means the dependencies need to be managed on the side of the test which is not ideal as well. I don't want you to change it here (maybe apart from ObjectMapper below), but I would like you to have an opinion and consider all pros and cons for the future :).

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/cmo/CMOReadyToSealEventHandlerTest.java [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+    @Mock
+    private ObjectMapper mapper;
+
+    @InjectMocks
+    private CMOReadyToSealEventHandler eventHandler;
+
+    @Test
+    void shouldSendNotificationForAdmin() {
+        CaseData caseData = caseData(false);
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/cmo/CMOReadyToSealEventHandlerTest.java:60]: As mentioned somewhere else I would probably expect two methods: (which can call this one with true / false). caseDataWithJudge and without. Such a small thing and increases the readability of the tests, now I need to jump below to understand what's going on.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/content/cmo/CMOReadyToSealContentProviderTest.java [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+import static uk.gov.hmcts.reform.fpl.enums.HearingType.CASE_MANAGEMENT;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.wrapElements;
+
+@ContextConfiguration(classes = {CMOReadyToSealContentProvider.class})
+class CMOReadyToSealContentProviderTest extends AbstractEmailContentProviderTest {
+    @Autowired
+    private CMOReadyToSealContentProvider contentProvider;
+
+    @MockBean
+    private EmailNotificationHelper notificationHelper;
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/content/cmo/CMOReadyToSealContentProviderTest.java:34]: Really same reason as above - I wouldn't mock it. With mocking you actually have more setup to do. If you add EmailNotificationHelper.class, TimeConfiguration.class to the ContextConfiguration, this test passes without this line and (lol @ BeforeEach tags someone...) 

Again I would like you to develop your own opinion, but I think mocking "helper" classes is not necessary. They are increasing the complexity (marginally), of test configuration, but don't increase the test maintainability.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+
+        data.put(NUM_HEARINGS_FIELD, numHearings);
+
+        return data;
+    }
+
+    public List<Element<HearingBooking>> getHearingsWithoutCMO(List<Element<HearingBooking>> hearings,
+                                                               List<Element<CaseManagementOrder>> unsealedOrders) {
+        return hearings.stream()
+            .filter(hearing -> associatedToReturnedCMO(hearing, unsealedOrders)
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java:85]: Is returning part of this story or just preparing for the future 🤔

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+    public Map<String, Object> getJudgeAndHearingDetails(UUID selectedHearing,
+                                                         List<Element<HearingBooking>> hearings) {
+        HearingBooking selected = getSelectedHearing(selectedHearing, hearings);
+        return Map.of(
+            CMO_JUDGE_INFO_FIELD, formatJudgeTitleAndName(selected.getJudgeAndLegalAdvisor()),
+            CMO_HEARING_INFO_FIELD, selected.toLabel(DATE)
+        );
+    }
+
+    public UUID mapToHearing(UUID selectedHearing, List<Element<HearingBooking>> hearings,
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java:99]: I don't like this method seems functional (mapToHreaing), but performs set, it's side-effect which I wouldn't expect here. Is the side effect required? If yes could you consider renaming and make it explicit by name what's really happening?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,427 @@
...
+        List<Element<HearingBooking>> hearings = List.of(element(hearing()), element(hearing()), element(hearing()));
+        UUID selected = hearings.get(2).getId();
+
+        HearingBooking selectedHearing = service.getSelectedHearing(selected, hearings);
+
+        assertThat(selectedHearing).isEqualTo(hearings.get(2).getValue());
+    }
+
+    @Test
+    void shouldThrowErrorWhenNoHearingMatchesTheIdPassed() {
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java:125]: Wouldn't PassedId read better?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,427 @@
...
+        Map<String, Object> expected = Map.of(
+            "pastHearingList", dynamicList(hearings.get(0).getId(), hearings.get(1).getId(), hearings.get(2).getId()),
+            "numHearings", "MULTI"
+        );
+
+        assertThat(initialPageData).isEqualTo(expected);
+    }
+
+    @Test
+    void shouldReturnCorrectPageDataWhenThereAreMultipleHearingsWithSomeAlreadyAssigned() {
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java:209]: Assigned? What assigned means here? Maybe something like AlreadyHavingCMO?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,427 @@
...
+            "cmoHearingInfo", "Send agreed CMO for Case management hearing, 1 February 2020."
+                + "\nThis must have been discussed by all parties at the hearing.",
+            "cmoJudgeInfo", "His Honour Judge Dredd"
+        );
+
+        assertThat(initialPageData).isEqualTo(expected);
+    }
+
+    @Test
+    void shouldReturnCorrectPageDataWhenThereIsOneRemainingHearingWithSomeAlreadyAssigned() {
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java:250]: Assigned seems a bit off for me, am I missing something? It's late so I might be!

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java [hunk 23]
```diff
@@ -0,0 +1,427 @@
...
+
+        Map<String, String> expected = Map.of(
+            "numHearings", "NONE"
+        );
+
+        assertThat(initialPageData).isEqualTo(expected);
+    }
+
+    @Test
+    void shouldReturnCorrectPageDataWhenThereAreNoRemainingHearingsWithoutCMOsAssigned() {
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java:283]: I am not sure if I like "shouldReturnCorrectPageData" of course the code should work. It's quite wordy but a bit obvious, would you consider rename?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java [hunk 24]
```diff
@@ -0,0 +1,427 @@
...
+    }
+
+    @Test
+    void shouldReturnJudgeNameAndTitleAndHearingInfo() {
+        List<Element<HearingBooking>> hearings = hearings();
+        Map<String, Object> details = service.getJudgeAndHearingDetails(hearings.get(0).getId(), hearings);
+
+        assertThat(details).isEqualTo(Map.of(
+            "cmoHearingInfo", "Case management hearing, 2 March 2020",
+            "cmoJudgeInfo", "His Honour Judge Dredd"
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java:348]: Executioner 💀

File: ccd-definition/CaseField/CareSupervision/CMO.json [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "multiHearingsWithCMOs",
+    "Label": "Hearings with CMOs already being reviewed",
+    "FieldType": "TextArea",
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "singleHearingsWithCMOs",
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/CaseField/CareSupervision/CMO.json:73]: Yeah can remove plurality, this was just a copy paste job changing the prefix

File: ccd-definition/CaseField/CareSupervision/CMO.json [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+    "Label": "Attach agreed draft CMO",
+    "HintText": "Files must be converted to PDF",
+    "FieldType": "Document",
+    "SecurityClassification": "Public",
+    "RegularExpression": ".pdf"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "numHearings",
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/CaseField/CareSupervision/CMO.json:49]: That names makes more sense will change, the variable was originally storing the number of hearings but it was messing up show hide conditions when 0 for some reason and then I forgot to change the name

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java (hunk 22 above)
Comment (lewisbirks) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java:250]: nah, names are hard and they probably could be clearer

Comment (lewisbirks) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java:250]: assigned in this case meant had a CMO Id attached I believe. But as you can see I am not sure what it meant a couple of days after and I wrote the test so it probably needs a rename

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadCMOSubmittedControllerTest.java [hunk 27]
```diff
@@ -0,0 +1,194 @@
...
+                eq(ADMIN_EMAIL),
+                anyMap(),
+                eq(String.valueOf(CASE_ID))
+            );
+
+            verifyNoMoreInteractions(notificationClient);
+        });
+    }
+
+    private CallbackRequest callbackRequest(boolean withAllocatedJudge) {
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadCMOSubmittedControllerTest.java:121]: Old cases I believe

[the remark this answers is not in the document]
Comment (lewisbirks) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadCMOSubmittedControllerTest.java:121]: These days it shouldn't be possible

File: ccd-definition/CaseField/CareSupervision/CMO.json [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+    "FieldType": "Collection",
+    "FieldTypeParameter": "UploadedCaseManagementOrder",
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "pastHearingList",
+    "Label": "Which CMO do you want to send to the judge to review?",
+    "HintText": "All CMOs must have been discussed by all parties at the hearing.",
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/CaseField/CareSupervision/CMO.json:16]: `hearingsWithoutCMOList` maybe?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java (hunk 7 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java:69]: Yeah was thinking that I had left too much in the controller

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java (hunk 10 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java:110]: Yeah, forgot to move this into the service, thanks for pointing it out

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java (hunk 18 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java:85]: it wasn't part of the story but if they return a CMO then the hearing wouldn't have been picked up in the list again and would have remained stuck in the draft cmo list

File: ccd-definition/ComplexTypes/CareSupervision/CaseManagementOrder/CaseManagementOrder.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (joez-f) [reply ccd-definition/ComplexTypes/CareSupervision/CaseManagementOrder/CaseManagementOrder.json:1]: Don't really need it, we already store a hearing string with the correctly formatted text and I think that is enough for me

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/HearingBooking.java (hunk 12 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/HearingBooking.java:58]: Fair enough, was wondering how generic to make it. Are you happy for me to leave it here? I think it is better than the alternative that was just merged in with notice of proceedings which has something very similar in the hearing booking service but happy to change if you think there is a better place.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/content/cmo/CMOReadyToSealContentProviderTest.java (hunk 17 above)
Comment (lewisbirks) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/email/content/cmo/CMOReadyToSealContentProviderTest.java:34]: I was trying to lessen the dependance on starting up spring here. Ideally `EmailNotificationHelper` wouldn't be a bean and #1433 is looking to resolve that. With this story I was trying out the idea that unit tests, unless really necessary, shouldn't start up spring as that could slow things down quite a bit relatively speaking.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadCMOSubmittedControllerTest.java [hunk 29]
```diff
@@ -0,0 +1,194 @@
...
+                eq(CMO_READY_FOR_JUDGE_REVIEW_NOTIFICATION_TEMPLATE_JUDGE),
+                eq(JUDGE_EMAIL),
+                anyMap(),
+                eq(String.valueOf(CASE_ID))
+            );
+        });
+    }
+
+    @Test
+    void shouldNotSendToJudgeIfAllocatedJudgeEmailNotProvided() {
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadCMOSubmittedControllerTest.java:104]: I thought I did have it in the event handler but it seems that I don't. Will move it there.

File: ccd-definition/CaseField/CareSupervision/CMO.json (hunk 25 above)
[the remark this answers is not in the document]
Comment (lewisbirks) [reply ccd-definition/CaseField/CareSupervision/CMO.json:73]: renamed, was a bad copy and paste job

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java (hunk 7 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java:69]: Probably the whole bit (including dynamicList) yes, although dynamicList is less important for me as it is "preparing the response"  I would consider something like that:

```suggestion
data.putAll(cmoService.prepareJudgeAndHearingDetails(dynamicList, caseData.getPastHearings(), caseDatta.getDraftUploadedCMOs))
```
?

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java:69]: Sounds good to me

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java (hunk 18 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java:85]: It wasn't too much effort, just adding some extra filtering and some tests to cover the scenario so I just did it

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java (hunk 20 above)
Comment (lewisbirks) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOServiceTest.java:125]: test no longer exists after service refactor

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java (hunk 19 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java:99]: refactored and renamed to `updateHearingWithCmoId`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java (hunk 18 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java:158]: 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/HearingBooking.java (hunk 12 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/HearingBooking.java:58]: Happy to leave it here for now.
